src/useField.js

```diff
@@ -0,0 +1,205 @@
+// @flow
+import * as React from 'react'
+import { fieldSubscriptionItems } from 'final-form'
+import flattenSubscription from './flattenSubscription'
+import type { FieldSubscription, FieldState, FormApi } from 'final-form'
+import type { UseFieldConfig, FieldInputProps, FieldRenderProps } from './types'
+import isReactNative from './isReactNative'
+import getValue from './getValue'
+import ReactFinalFormContext from './context'
+
+const all: FieldSubscription = fieldSubscriptionItems.reduce((result, key) => {
+  result[key] = true
+  return result
+}, {})
+
+export const defaultFormat = (value: ?any, name: string) =>
+  value === undefined ? '' : value
+export const defaultParse = (value: ?any, name: string) =>
+  value === '' ? undefined : value
+
+const useField = (
+  name: string,
+  {
+    allowNull,
+    component,
+    defaultValue,
+    format,
+    formatOnBlur,
+    initialValue,
+    isEqual,
+    multiple,
+    parse,
+    subscription,
+    type,
+    validate,
+    validateFields,
+    value: _value
+  }: UseFieldConfig = {}
+): FieldRenderProps => {
+  const reactFinalForm: ?FormApi = React.useContext(ReactFinalFormContext)
+  // istanbul ignore next
+  if (process.env.NODE_ENV !== 'production' && !reactFinalForm) {
+    console.error(
+      'Warning: useField must be used inside of a ReactFinalForm component'
+    )
+  }
+
+  // keep ref to most recent copy of validate function
+  const validateRef = React.useRef(validate)
+  validateRef.current = validate
```

Comment:
IMHO this is not really the proper way of saving latest value, it's slightly better to use such logic as here https://github.com/Andarist/use-latest/ 

the difference is subtle and probably not really relevant, but I believe the other way of doing this is more predictable - gonna come back with explanation 2morrow

Reply:
Consider what might in theory happen - react might render at arbitrary times (and thus updating this ref) and in theory `validate` might change each time, so when `validate` get called it's one of those functions, but we have no idea which one.

If we update it in useEffect (after commit phase) then we at least know which version of that function it is, the last one from before commit phase.

Moving it to useEffect also corresponds nicely with how classes work, what you want here is `this.props.validate`, but because of heavy closure usage in hooks it's not as simple as reaching `props.validate`, but conceptually this is what you want, right? So in class worls `this.props` gets updated during commit phase, kinda just before componentDidUpdate.

Reply:
Gotcha! Great explanation! 👍